File: Day09/C++/que4linkedlist_middle.cpp
#include <iostream>
using namespace std;

class SinglyList
{
private:
    class Node
    {
    public:
        int data;
        Node *next;
        Node(int val) : data(val), next(nullptr) {}
    };

    Node *head;

public:
    SinglyList() : head(nullptr) {}

    void display()
    {
        cout << "List: ";
        Node *trav = head;
        while (trav != nullptr)
        {
            cout << trav->data << " -> ";
            trav = trav->next;
        }
        cout << endl;
    }

    void addLast(int val)
    {
        Node *newNode = new Node(val);
        if (head == nullptr)
        {
            head = newNode;
        }
        else
        {
            Node *trav = head;
            while (trav->next != nullptr)
            {
                trav = trav->next;
            }
            trav->next = newNode;
        }
    }

    void printMid()
    {
        if (head != nullptr)
        {
            Node *slow = head;
            Node *fast = head;
            while (fast != nullptr && fast->next != nullptr)
            {
                slow = slow->next;
                fast = fast->next->next;
            }
            cout << "Middle Node: " << slow->data << endl;
        }
        else
        {
            cout << "List is empty!" << endl;
        }
    }

    ~SinglyList()
    {
        Node *trav = head;
        while (trav != nullptr)
        {
            Node *temp = trav;
            trav = trav->next;
            delete temp;
        }
    }
};

int main()
{
    SinglyList list;
    list.addLast(10);
    list.addLast(20);
    list.addLast(30);
    list.addLast(40);
    list.addLast(50);
    list.addLast(60);
    list.addLast(70);
    list.addLast(80);
    list.display();
    list.printMid();
    return 0;
}

File: Day09/Python/que4linkedlist-middle.py
class SinglyList:
    class Node:
        def __init__(self, val):
            self.data = val
            self.next = None

    def __init__(self):
        self.head = None

    def display(self):
        print("List: ", end="")
        trav = self.head
        while trav is not None:
            print(trav.data, "-> ", end="")
            trav = trav.next
        print()

    def addLast(self, val):
        newNode = self.Node(val)
        if self.head is None:
            self.head = newNode
        else:
            trav = self.head
            while trav.next is not None:
                trav = trav.next
            trav.next = newNode

    def printMid(self):
        if self.head is not None:
            slow = self.head
            fast = self.head
            while fast is not None and fast.next is not None:
                slow = slow.next
                fast = fast.next.next
            print("Middle Node:", slow.data)
        else:
            print("List is empty!")


def main():
    list = SinglyList()
    list.addLast(10)
    list.addLast(20)
    list.addLast(30)
    list.addLast(40)
    list.addLast(50)
    list.addLast(60)
    list.addLast(70)
    list.addLast(80)
    list.display()
    list.printMid()


if __name__ == "__main__":
    main()
